app/models/region.rb

```diff
@@ -0,0 +1,51 @@
+class Region < ApplicationRecord
+  ltree :path
+  extend FriendlyId
+  friendly_id :name, use: :slugged
+
+  validates :name, presence: true
+  validates :slug, presence: true, uniqueness: true
+  validates :path, presence: true
+
+  belongs_to :kind, class_name: "RegionKind", foreign_key: "region_kind_id"
+  belongs_to :source, polymorphic: true, optional: true
+
+  before_discard do
+    self.path = nil
+  end
+
+  def self.create_region_from(parent:, kind:, name: nil, source: nil)
+    raise ArgumentError, "Provide either a name or a source" if (name && source) || (name.blank? && source.blank?)
+    region_name = name || source.name
+    region = Region.new name: region_name, kind: kind
+    region.send :set_slug
+    region.source = source if source
+    region.path = "#{parent.path}.#{region.slug.tr("-", "_")}"
+    region.save!
+    region
+  end
+
+  def self.backfill!
+    instance_kind = RegionKind.find_by!(name: "Instance")
```

Comment:
Can we use something like `Root` or `RootRegion` instead of `Instance` for these top level `RegionKinds`? `Instance` doesn't really convey anything to me semantically.

Reply:
Yeah we went back and forth on the name for the top level thing.  I'll change it to `Root`.